src/net_processing.cpp

```diff
@@ -380,8 +380,18 @@ struct CNodeState {
     const CBlockIndex* pindexBestKnownBlock{nullptr};
     //! The hash of the last unknown block this peer has announced.
     uint256 hashLastUnknownBlock{};
-    //! The last full block we both have.
-    const CBlockIndex* pindexLastCommonBlock{nullptr};
+
+    //! The last full block we both have (per chainstate).
+    //!
+    //! This is namespaced by chainstate to allow syncing two separate chainstates
+    //! simultaneously from a single peer.
+    //!
+    //! Nota bene: this is contingent on the ChainstateManager not destructing
+    //! any CChainState objects which were in use at any point (e.g. a background
+    //! validation chainstate which has completed) until the end of
+    //! init.cpp:Shutdown(), else we'll have bad pointers here.
+    std::map<const CChainState*, const CBlockIndex*> chainstate_to_last_common_block = {};
```

Comment:
Having a one or two element map for every peer seems a bit icky.

Is this even necessary though? If you have a common snapshot block, then your common IBD block will be whatever your IBD tip is, no? And if you do have a snapshot block, but don't have a common snapshot block with this peer, don't you want to disconnect?

I guess I'm suggesting replacing this by a `CBlockIndex*` to trace the last common block, and a `bool is_last_common_block_from_snapshot`?

Reply:
> The consensus has been that we want as few parts of the system as possible knowing about "snapshots" per se.

Sure; but the map already means the code needs to know about snapshots per se.

> if we were ever to further parallelize validation with more than one background validation chainstate, we could keep this structure.

That seems crazy to me; managing two utxo sets is already a lot of space, increasing that is costly, and I don't think validating the chain at many points would help with any bottlenecks.

But you shouldn't even need the bool; the fact that the snapshots are all at different points in the same chain should let you just say: `last_common_block_for_this_chainstate = min(tip, last_common_block)` when picking where to start, and `if (last_common_block->nHeight < pindex->nHeight) last_common_block = pindex` when updating.

<details>
<summary>(click for suggestion)</summary>

```diff
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -391,7 +391,7 @@ struct CNodeState {
     //! any CChainState objects which were in use at any point (e.g. a background
     //! validation chainstate which has completed) until the end of
     //! init.cpp:Shutdown(), else we'll have bad pointers here.
-    std::map<const CChainState*, const CBlockIndex*> chainstate_to_last_common_block = {};
+    const CBlockIndex* last_common_block = nullptr;
 
     //! The best header we have sent our peer.
     const CBlockIndex* pindexBestHeaderSent{nullptr};
@@ -757,7 +757,7 @@ private:
 
     bool TipMayBeStale() EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
-    /** Update chainstate_to_last_common_block and add not-in-flight missing successors
+    /** Update last_common_block and add not-in-flight missing successors
      * to vBlocks, until it has at most count entries.
      */
     void FindNextBlocksToDownload(
@@ -1140,6 +1140,13 @@ void PeerManagerImpl::UpdateBlockAvailability(NodeId nodeid, const uint256 &hash
     }
 }
 
+static const CBlockIndex* GetLastCommonBlock(const CBlockIndex* tip, const CBlockIndex* last_common_block) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
+{
+    if (last_common_block == nullptr || tip == nullptr) return nullptr;
+    if (tip->nHeight <= last_common_block->nHeight) return tip;
+    return last_common_block;
+}
+
 void PeerManagerImpl::FindNextBlocksToDownload(
     const CChainState* const chainstate,
     NodeId nodeid,
@@ -1167,27 +1174,22 @@ void PeerManagerImpl::FindNextBlocksToDownload(
         return;
     }
 
-    if (!state->chainstate_to_last_common_block.count(chainstate)) {
+    if (!state->last_common_block) {
         // Bootstrap quickly by guessing a parent of our best tip is the forking point.
         // Guessing wrong in either direction is not a problem.
-        //
-        // Namespace this by chainstate so that we can simultaneously sync two
-        // separate chainstates at different heights.
-        state->chainstate_to_last_common_block[chainstate] = our_chain[
-            std::min(state->pindexBestKnownBlock->nHeight, our_chain.Height())];
+        state->last_common_block = our_chain[std::min(state->pindexBestKnownBlock->nHeight, our_chain.Height())];
     }
 
     // If the peer reorganized, our previous chainstate_to_last_common_block may not be an ancestor
     // of its current tip anymore. Go back enough to fix that.
-    state->chainstate_to_last_common_block[chainstate] = LastCommonAncestor(
-        state->chainstate_to_last_common_block[chainstate], state->pindexBestKnownBlock);
+    state->last_common_block = LastCommonAncestor(state->last_common_block, state->pindexBestKnownBlock);
 
-    if (state->chainstate_to_last_common_block[chainstate] == state->pindexBestKnownBlock) {
+    if (state->last_common_block == state->pindexBestKnownBlock) {
         return;
     }
 
     std::vector<const CBlockIndex*> vToFetch;
-    const CBlockIndex *pindexWalk = state->chainstate_to_last_common_block[chainstate];
+    const CBlockIndex* pindexWalk = GetLastCommonBlock(our_tip, state->last_common_block);
     // Never fetch further than the best block we know the peer has, or more than BLOCK_DOWNLOAD_WINDOW + 1 beyond the last
     // linked block we have in common with this peer. The +1 is so we can detect stalling, namely if we would be able to
     // download that next block if the window were 1 larger.
@@ -1209,7 +1211,7 @@ void PeerManagerImpl::FindNextBlocksToDownload(
 
         // Iterate over those blocks in vToFetch (in forward direction), adding the ones that
         // are not yet downloaded and not in flight to vBlocks. In the meantime, update
-        // chainstate_to_last_common_block as long as all ancestors are already downloaded, or if it's
+        // last_common_block as long as all ancestors are already downloaded, or if it's
         // already part of our chain (and therefore don't need it even if pruned).
         for (const CBlockIndex* pindex : vToFetch) {
             if (!pindex->IsValid(BLOCK_VALID_TREE)) {
@@ -1221,8 +1223,11 @@ void PeerManagerImpl::FindNextBlocksToDownload(
                 return;
             }
             if (pindex->nStatus & BLOCK_HAVE_DATA || our_chain.Contains(pindex)) {
-                if (pindex->HaveTxsDownloaded())
-                    state->chainstate_to_last_common_block[chainstate] = pindex;
+                if (pindex->HaveTxsDownloaded()) {
+                    if (state->last_common_block == nullptr || pindex->nHeight > state->last_common_block->nHeight) {
+                        state->last_common_block = pindex;
+                    }
+                }
             } else if (!IsBlockRequested(pindex->GetBlockHash())) {
                 // The block is not already downloaded, and not yet in flight.
                 if (pindex->nHeight > nWindowEnd) {
@@ -1434,13 +1439,8 @@ bool PeerManagerImpl::GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) c
         stats.nSyncHeight = state->pindexBestKnownBlock ? state->pindexBestKnownBlock->nHeight : -1;
         stats.nCommonHeight = -1;
 
-        // Take the max common height with this peer across chainstates.
-        for (const auto& it : state->chainstate_to_last_common_block) {
-            const CBlockIndex* block = it.second;
-            if (block && block->nHeight > stats.nCommonHeight) {
-                stats.nCommonHeight = block->nHeight;
-            }
-        }
+        if (state->last_common_block) stats.nCommonHeight = state->last_common_block->nHeight;
+
         for (const QueuedBlock& queue : state->vBlocksInFlight) {
             if (queue.pindex)
                 stats.vHeightInFlight.push_back(queue.pindex->nHeight);
```
</details>